sklearn/impute.py

```diff
@@ -873,3 +880,309 @@ def fit(self, X, y=None):
+            donor_row_idx_ravel = donors_idx.ravel()
+
+            # Retrieve donor values and calculate kNN score
+            fitted_X_temp = fitted_X[pdonors_row_idx]
+            donors = fitted_X_temp[donor_row_idx_ravel, c].reshape(
+                (-1, self.n_neighbors))
+            donors_mask = _get_mask(donors, self.missing_values)
+            donors = np.ma.array(donors, mask=donors_mask)
+
+            # Final imputation
+            imputed = np.ma.average(donors, axis=1,
+                                    weights=weight_matrix)
+            X[receivers_row_idx, c] = imputed.data
+        return X
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+
+        # Check data integrity and calling arguments
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        if not force_all_finite:
+            if self.metric not in _MASKED_METRICS and not callable(
+                    self.metric):
+                raise ValueError(
+                    "The selected metric does not support NaN values.")
+        X = check_array(X, accept_sparse=False, dtype=np.float64,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Check if % missing in any column > col_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.col_max_missing)):
+            raise ValueError("Some column(s) have more than {}% missing values"
+                             .format(self.col_max_missing * 100))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > row_max_missing
+        bad_rows = mask.sum(axis=1) > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. These "
+                "rows are not included as donor neighbors."
+                    .format(self.row_max_missing * 100))
+
+            # Remove rows that have more than row_max_missing % missing
+            X = X[~bad_rows, :]
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
+            raise ValueError("There are only %d samples, but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+        self.fitted_X_ = X
```

Comment:
please call this `fit_X_` (i.e. the X when fit was called; there is nothing "fitted" about X itself)